src/github.com/matrix-org/dendrite/federationsender/storage/storage.go

```diff
@@ -62,30 +62,47 @@ func (d *Database) prepare() error {
 }
 
 // UpdateRoom updates the joined hosts for a room and returns what the joined
-// hosts were before the update.
+// hosts were before the update, or nil if this was a duplicate message.
+// This is called when we receive a message from kafka, so we pass in
+// oldEventID and newEventID to check that we haven't missed any messages or
+// this isn't a duplicate message.
 func (d *Database) UpdateRoom(
 	ctx context.Context,
 	roomID, oldEventID, newEventID string,
 	addHosts []types.JoinedHost,
 	removeHosts []string,
-) (joinedHosts []types.JoinedHost, err error) {
-	err = common.WithTransaction(d.db, func(txn *sql.Tx) error {
-		if err = d.insertRoom(ctx, txn, roomID); err != nil {
+) (*[]types.JoinedHost, error) {
+	var joinedHostsPtr *[]types.JoinedHost
+
+	err := common.WithTransaction(d.db, func(txn *sql.Tx) error {
+		err := d.insertRoom(ctx, txn, roomID)
+		if err != nil {
 			return err
 		}
+
 		lastSentEventID, err := d.selectRoomForUpdate(ctx, txn, roomID)
 		if err != nil {
 			return err
 		}
+
+		if lastSentEventID == newEventID {
```

Comment:
why only a strict equality check? what if we processed more than one message?

Reply:
We can only get consecutive duplicate messages from kafka (when we crash before persisting our stream position), and I'm not sure we really want to be storing more event IDs.

Reply:
(I've added to the comment beneath it to explain)